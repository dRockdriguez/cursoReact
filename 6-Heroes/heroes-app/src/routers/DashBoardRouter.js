import React from "react";
import { Routes, Route } from "react-router-dom";
import { MarvelScreen } from "../components/marvel/MarvelScreen";
import { DcScreen } from "../components/dc/DcScreen";
import { SearchScreen } from "../components/search/SearchScreen";
import { Navbar } from "../components/ui/NavBar";
import { HeroeScreen } from "../components/heroes/HeroeScreen";

export const DashBoardRouter = () => {
  return (
    <>
      <Navbar />
      <div className="container mt-5">
        <Routes>
          <Route path="marvel" element={<MarvelScreen />} />
          <Route path="dc" element={<DcScreen />} />
          <Route path="search" element={<SearchScreen />} />
          <Route path=":publisher/heroe/:id" element={<HeroeScreen />} />
        </Routes>
      </div>
    </>
  );
};
