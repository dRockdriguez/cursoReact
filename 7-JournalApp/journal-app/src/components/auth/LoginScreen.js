import React from "react";
import { Link } from "react-router-dom";

export const LoginScreen = () => {
  return (
    <>
      <h3 className="auth__title mb-2">Login</h3>

      <form>
        <input
          className="auth__input mb-1"
          type="email"
          placeholder="Email..."
          name="email"
          autoComplete="off"
        />
        
        <input
          className="auth__input mb-1"
          type="password"
          placeholder="Password..."
          name="password"
        />

        <button className="btn btn-primary btn-block" type="submit">
          Go
        </button>

        <div className="auth__social-networks">
          <p className="mb-2">Login with Social Networks</p>

          <div className="google-btn">
            <div className="google-icon-wrapper">
              <img
                className="google-icon"
                src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
                alt="google button"
              />
            </div>
            <p className="btn-text">
              <b>Sign in with google</b>
            </p>
          </div>

          <Link className="link mt-2" to="/auth/register">
            Create new account
          </Link>
        </div>
      </form>
    </>
  );
};
