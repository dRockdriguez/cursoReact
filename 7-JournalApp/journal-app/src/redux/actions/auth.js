import Swal from "sweetalert2";
import { types } from "../types/types";
import { googleAuthProvider } from "../../firebase/firebase-config";
import {
  getAuth,
  signInWithPopup,
  createUserWithEmailAndPassword,
  updateProfile,
  signInWithEmailAndPassword,
  signOut,
} from "firebase/auth";
import { finishLoading, startLoading } from "./ui";

export const login = (uid, displayName) => ({
  type: types.login,
  payload: {
    uid,
    displayName,
  },
});

export const startLoginEmailPassword = (email, password) => {
  return async (dispatch) => {
    try {
      dispatch(startLoading());
      const auth = getAuth();
      const { user } = await signInWithEmailAndPassword(auth, email, password);
      const { uid, displayName } = user;
      dispatch(login(uid, displayName));
    } catch (e) {
      showError(e.message);
    } finally {
      dispatch(finishLoading());
    }
  };
};

export const startGoogleLogin = () => {
  return async (dispatch) => {
    try {
      dispatch(startLoading());
      const auth = getAuth();
      const {
        user: { uid, displayName },
      } = await signInWithPopup(auth, googleAuthProvider);
      dispatch(login(uid, displayName));
    } catch (e) {
      showError(e.message);
    } finally {
      dispatch(finishLoading());
    }
  };
};

export const startRegister = ({ email, password, name }) => {
  return async (dispatch) => {
    try {
      dispatch(startLoading());
      const auth = getAuth();
      const { user } = await createUserWithEmailAndPassword(
        auth,
        email,
        password
      );
      await updateProfile(user, { displayName: name });
      const { uid, displayName } = user;
      dispatch(login(uid, displayName));
    } catch (e) {
      showError(e.message);
    } finally {
      dispatch(finishLoading());
    }
  };
};

export const startLogout = () => {
  return async (dispatch) => {
    try {
      dispatch(startLoading());
      const auth = getAuth();
      await signOut(auth);
      dispatch(logout());
    } catch (e) {
      showError(e.message);
    } finally {
      dispatch(finishLoading());
    }
  };
};

export const logout = () => ({
  type: types.logout,
});

const showError = (message) => {
  Swal.fire({
    icon: "error",
    title: "Oops...",
    text: message,
  });
};
