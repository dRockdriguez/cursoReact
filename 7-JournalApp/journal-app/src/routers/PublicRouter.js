import React from "react";
import { Navigate } from "react-router-dom";

export const PublicRouter = ({ isAuth, children }) => {
  return isAuth ? <Navigate to="/" /> : children;
};
