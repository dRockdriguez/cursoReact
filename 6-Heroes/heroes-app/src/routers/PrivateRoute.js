import { useContext } from "react";
import { AuthContext } from "../auth/AuthContext";
import { Navigate } from "react-router-dom";

export const PrivateRoute = ({ children }) => {
  const { user } = useContext(AuthContext);
  return user.logged ? children : <Navigate to="login" />;
};
