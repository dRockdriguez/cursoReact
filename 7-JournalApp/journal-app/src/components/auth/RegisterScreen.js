import React from "react";
import { useDispatch } from "react-redux";
import { Link } from "react-router-dom";
import { useForm } from "../../hooks/useForm";
import validator from "validator";
import { setError, removeError } from "../../redux/actions/ui";

export const RegisterScreen = () => {
  const dispatch = useDispatch();

  const [formValues, handleInputChange] = useForm({
    name: "",
    email: "",
    password: "",
    confirm: "",
  });

  const { name, email, password, confirm } = formValues;

  const handleRegister = (e) => {
    e.preventDefault();

    const valid = isValidForm();
    if (valid.valid) {
      dispatch(removeError());
    } else {
      dispatch(setError(valid.error));
    }
  };

  const isValidForm = () => {
    if (name.trim().length <= 0) {
      return {
        valid: false,
        error: "The name is invalid",
      };
    } else if (!validator.isEmail(email)) {
      return {
        valid: false,
        error: "The email is invalid",
      };
    } else if (password.length <= 5 || password !== confirm) {
      return {
        valid: false,
        error: "The password is invalid",
      };
    }
    return {
      valid: true,
    };
  };

  return (
    <>
      <h3 className="auth__title mb-2">Sign up</h3>

      <form onSubmit={handleRegister}>
        <div className="auth__alert-error">Hola</div>

        <input
          className="auth__input mb-1"
          type="name"
          placeholder="Name..."
          name="name"
          autoComplete="off"
          value={name}
          onChange={handleInputChange}
        />

        <input
          className="auth__input mb-1"
          type="email"
          placeholder="Email..."
          name="email"
          autoComplete="off"
          value={email}
          onChange={handleInputChange}
        />

        <input
          className="auth__input mb-1"
          type="password"
          placeholder="Password..."
          name="password"
          value={password}
          onChange={handleInputChange}
        />

        <input
          className="auth__input mb-1"
          type="password"
          placeholder="Confirm password..."
          name="confirm"
          value={confirm}
          onChange={handleInputChange}
        />

        <button className="btn btn-primary btn-block mb-2" type="submit">
          Go
        </button>

        <Link className="link" to="/auth/login">
          Already registered?
        </Link>
      </form>
    </>
  );
};
