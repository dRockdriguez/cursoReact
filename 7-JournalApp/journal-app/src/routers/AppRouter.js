import React, { useEffect, useState } from "react";
import { Routes, Route, BrowserRouter, Navigate } from "react-router-dom";
import { AuthRouter } from "./AuthRouter";
import { JournalScreen } from "../components/journal/JournalScreen";
import { useDispatch, useSelector } from "react-redux";
import { selectLoading } from "../redux/selectors/ui";
import { Loader } from "../components/ui/Loader";
import { getAuth, onAuthStateChanged } from "firebase/auth";
import { login } from "../redux/actions/auth";
import { PublicRouter } from "./PublicRouter";
import { PrivateRouter } from "./PrivateRouter";

export const AppRouter = () => {
  const loading = useSelector(selectLoading);
  const [checking, setChecking] = useState(true);
  const [isLogged, setIsLogged] = useState(false);

  const dispatch = useDispatch();

  useEffect(() => {
    const auth = getAuth();
    onAuthStateChanged(auth, (user) => {
      if (user?.uid) {
        dispatch(login(user.uid, user.displayName));
        setIsLogged(true);
      } else {
        setIsLogged(false);
      }
      setChecking(false);
    });
  }, [dispatch, setChecking, setIsLogged]);

  return (
    <>
      {(checking || loading) && <Loader />}
      <BrowserRouter>
        <Routes>
          <Route
            path="/auth/*"
            element={
              <PublicRouter isAuth={isLogged}>
                <AuthRouter />
              </PublicRouter>
            }
          />
          <Route
            path="/"
            element={
              <PrivateRouter isAuth={isLogged}>
                <JournalScreen />
              </PrivateRouter>
            }
          />
          <Route path="*" element={<Navigate to="/auth/login" />} />
        </Routes>
      </BrowserRouter>
    </>
  );
};
