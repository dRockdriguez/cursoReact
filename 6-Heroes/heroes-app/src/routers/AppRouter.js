import React from "react";
import { Routes, Route, BrowserRouter } from "react-router-dom";
import { LoginScreen } from "../components/login/LoginScreen";
import { DashBoardRouter } from "./DashBoardRouter";
import { PrivateRoute } from "./PrivateRoute";
import { PublicRoute } from "./PublicRoute";

export const AppRouter = () => {
  return (
    <BrowserRouter>
      <Routes>
        <Route
          path="/login"
          element={
            <PublicRoute>
              <LoginScreen />
            </PublicRoute>
          }
        />
        <Route
          path="/*"
          element={
            <PrivateRoute>
              <DashBoardRouter />
            </PrivateRoute>
          }
        />
      </Routes>
    </BrowserRouter>
  );
};
