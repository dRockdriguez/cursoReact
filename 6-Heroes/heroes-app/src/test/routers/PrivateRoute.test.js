import { PrivateRoute } from "../../routers/PrivateRoute";
import { mount } from "enzyme";
import { AuthContext } from "../../auth/AuthContext";
import { MemoryRouter } from "react-router-dom";

jest.mock("react-router-dom", () => ({
    ...jest.requireActual("react-router-dom"),
    Navigate: () => <h1>Login</h1>
}));

describe("PrivateRoute", () => {
  test("Debe mostrarse el componente si está auntenticado y guardar en localstorage", () => {
    Storage.prototype.setItem = jest.fn();

    const context = {
      user: {
        logged: true,
        name: "David",
      },
    };

    const wrapper = mount(
      <AuthContext.Provider value={context}>
        <MemoryRouter initialEntries={["/"]}>
          <PrivateRoute>
              <h1>Private</h1>
          </PrivateRoute>
        </MemoryRouter>
      </AuthContext.Provider>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find("h1").text().trim()).toBe("Private");
    expect(localStorage.setItem).toHaveBeenCalledWith("lastPath", "//");
  });

  test("Debe ir al login si no está autenticado", () => {
    Storage.prototype.setItem = jest.fn();

    const context = {
      user: {
        logged: false,
      },
    };

    const wrapper = mount(
      <AuthContext.Provider value={context}>
        <MemoryRouter initialEntries={["/"]}>
          <PrivateRoute>
              <h1>Private</h1>
          </PrivateRoute>
        </MemoryRouter>
      </AuthContext.Provider>
    );
    expect(wrapper.find("h1").text().trim()).toBe("Login");
  });
});
