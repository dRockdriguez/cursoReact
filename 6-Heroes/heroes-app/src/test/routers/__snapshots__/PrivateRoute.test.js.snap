// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrivateRoute Debe mostrarse el componente si est√° auntenticado y guardar en localstorage 1`] = `
<h1>
  Private
</h1>
`;
